src/hwc/compile/luna/luna.go

```diff
@@ -0,0 +1,238 @@
+	BuildDir               string
+	LunaDir                string
+	credentials            *LunaCredentials
+	ClientPrivateKeyPath   string
+	ClientCertificatePath  string
+	ServerCertificatesPath string
+}
+
+type Client struct {
+	Certificate string
+	PrivateKey  string `json:"private-key"`
+}
+
+type Server struct {
+	Name        string
+	Certificate string
+}
+
+type Group struct {
+	Label   string
+	Members []string
+}
+
+type LunaCredentials struct {
+	Client  Client
+	Servers []Server
+	Groups  []Group
+}
+
+func NewLuna(log *libbuildpack.Logger, buildDir string) *Luna {
+	l := new(Luna)
+	l.Log = log
+	l.BuildDir = buildDir
+	l.LunaDir = filepath.Join(l.BuildDir, ".cloudfoundry", ".luna")
+	l.ClientPrivateKeyPath = filepath.Join(l.LunaDir, "clientPrivateKey.pem")
+	l.ClientCertificatePath = filepath.Join(l.LunaDir, "clientCertificate.pem")
+	l.ServerCertificatesPath = filepath.Join(l.LunaDir, "serverCertificates.pem")
+	return l
+}
+
+func validateCredentials(credentials *LunaCredentials) bool {
+	if credentials.Client.Certificate == "" || credentials.Client.PrivateKey == "" {
+		return false
+	}
+	if len(credentials.Servers) == 0 {
+		return false
+	}
+	for _, server := range credentials.Servers {
+		if server.Name == "" || server.Certificate == "" {
+			return false
+		}
+	}
+	for _, group := range credentials.Groups {
+		if group.Label == "" || len(group.Members) == 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func (l *Luna) parseVCAPServices() {
+	vcapServices := os.Getenv("VCAP_SERVICES")
+	var f map[string]interface{}
+	json.Unmarshal([]byte(vcapServices), &f)
+	for key, _ := range f {
+		services, ok := f[key].([]interface{})
+		if !ok {
+			continue
+		}
+		for _, v1 := range services {
+			creds1, ok := v1.(map[string]interface{})
+			if !ok {
+				continue
+			}
+			creds2, ok := creds1["credentials"]
+			if !ok {
+				continue
+			}
+			credentialsJson, err := json.Marshal(creds2)
+			if err != nil {
+				continue
+			}
+			var credentials LunaCredentials
+			err = json.Unmarshal(credentialsJson, &credentials)
+			if err == nil && validateCredentials(&credentials) {
+				l.credentials = &credentials
+				return
+			}
+		}
+	}
+}
+
+func (l *Luna) createLunaDirectory() error {
+	return os.MkdirAll(l.LunaDir, 0755)
+}
+
+func (l *Luna) writeNTLSCredentials() error {
+
+	privateKey := []byte(l.credentials.Client.PrivateKey)
+	err := ioutil.WriteFile(l.ClientPrivateKeyPath, privateKey, 0640)
+	if err != nil {
+		return err
+	}
+
+	clientCertificate := []byte(l.credentials.Client.Certificate)
+	err = ioutil.WriteFile(l.ClientCertificatePath, clientCertificate, 0644)
+	if err != nil {
+		return err
+	}
+
+	var serverCertificates string
+	serverCertificates = ""
+	for _, server := range l.credentials.Servers {
+		serverCertificates += server.Certificate
+	}
+	err = ioutil.WriteFile(l.ServerCertificatesPath, []byte(serverCertificates), 0644)
+
+	return err
+}
+
+func (l *Luna) writeCrystokiIni() error {
```

Comment:
This function is a bit hard to read. Because the goal is to output a single file, this seems like a good place to use a templating language. In particular, using Go's [`text/template`](https://golang.org/pkg/text/template/) seems like a good option.

Reply:
Good suggestion.  I will modify the code to use ```text/template```.